File: python/secret-handshake/secret_handshake.py
KEYWORDS = ['wink', 'double blink', 'close your eyes', 'jump', 'reverse']
encode_table = {2 ** i:KEYWORDS[i] for i in range(len(KEYWORDS))}
decode_table = {v:k for k, v in encode_table.items()}

def handshake(number):
    if isinstance(number, str):
        try:
            number = int(number, 2)
        except:
            return []

    if number < 0:
        return []

    dl = []
    for d, word in encode_table.items():
        if number & d:
            if word == 'reverse':
                dl.reverse()
            else:
                dl.append(word)
    return dl

def code(l):
    dl = [decode_table[word] for word in l if word in decode_table]
    if len(dl) != len(l):
        return '0'

    if sorted(dl) != dl:
        dl.append(16)
        dl.reverse()

    return bin(sum(dl))[2:]

File: ruby/secret-handshake/secret_handshake.rb
class SecretHandshake
  HANDSHAKE = ['wink', 'double blink', 'close your eyes', 'jump'].freeze

  attr_reader :commands

  def initialize(number)
    @commands = []
    return unless number.is_a? Integer
    HANDSHAKE.each do |shake|
      @commands << shake if marked?(number)
      number >>= 1
    end
    @commands.reverse! if marked?(number)
  end

  private

  def marked?(number)
    number % 2 == 1
  end
end
